File: backend/anomaly_backend/routes/simulation.py
from typing import Annotated, cast

from fastapi import APIRouter, Depends, Query
from sqlalchemy.engine import RowMapping
from sqlalchemy.ext.asyncio import AsyncConnection

from anomaly_backend.contracts import (
    CorpusDeviceId,
    ScopeMetricsModel,
    SetSimActiveModelRequest,
    SetSimActiveModelResponse,
    SimAlertEventModel,
    SimMetricsResponse,
    SimModel,
    SimModelsResponse,
)
from anomaly_backend.db import get_connection
from anomaly_backend.evaluation import ScopeMetrics
from anomaly_backend.problems import NotFound, new_request_id
from anomaly_backend.sim_metrics import assemble_sim_metrics
from anomaly_backend.sql import set_sim_active_model, sim_metrics_source, sim_model_rows


router = APIRouter()

_SIM_DEVICE_ID: CorpusDeviceId = "b02f3872-simulasi-injeksi"


@router.get("/api/simulation/models", response_model=SimModelsResponse)
async def simulation_models(
    connection: Annotated[AsyncConnection, Depends(get_connection)],
) -> SimModelsResponse:
    rows = await sim_model_rows(connection, _SIM_DEVICE_ID)
    return SimModelsResponse(
        request_id=new_request_id(),
        device_id=_SIM_DEVICE_ID,
        models=[_model(row) for row in rows],
    )


@router.post("/api/simulation/active-model", response_model=SetSimActiveModelResponse)
async def set_active_model(
    request: SetSimActiveModelRequest,
    connection: Annotated[AsyncConnection, Depends(get_connection)],
) -> SetSimActiveModelResponse:
    active = await set_sim_active_model(
        connection, device_id=_SIM_DEVICE_ID, model_version=request.model_version
    )
    if active is None:
        raise NotFound("The artifact model was not found for the simulation device")
    return SetSimActiveModelResponse(
        request_id=new_request_id(),
        device_id=_SIM_DEVICE_ID,
        active_model_version=active,
    )


@router.get("/api/simulation/metrics", response_model=SimMetricsResponse)
async def simulation_metrics(
    model_version: Annotated[str, Query(min_length=1)],
    connection: Annotated[AsyncConnection, Depends(get_connection)],
    cooldown_samples: Annotated[int, Query(ge=1)] = 10,
) -> SimMetricsResponse:
    source = await sim_metrics_source(
        connection, device_id=_SIM_DEVICE_ID, model_version=model_version
    )
    if source is None:
        raise NotFound("The artifact model was not found for the simulation device")
    if not source["window_rows"]:
        raise NotFound("No replay results exist for this model on the simulation device")
    metrics = assemble_sim_metrics(
        model_version=model_version,
        threshold=cast(float, source["threshold"]),
        window_size=cast(int, source["window_size"]),
        frame_count=cast(int, source["frame_count"]),
        window_rows=cast(list, source["window_rows"]),
        event_rows=cast(list, source["event_rows"]),
        segment_rows=cast(list, source["segment_rows"]),
        cooldown_samples=cooldown_samples,
    )
    return SimMetricsResponse(
        request_id=new_request_id(),
        device_id=_SIM_DEVICE_ID,
        model_version=metrics.model_version,
        threshold=metrics.threshold,
        window_size=metrics.window_size,
        frame_count=metrics.frame_count,
        event_count=metrics.event_count,
        scored_windows=metrics.scored_windows,
        timestamp_scope=_scope(metrics.timestamp),
        overlapping_scope=_scope(metrics.overlapping),
        bins_scope=_scope(metrics.bins),
        operational_event_count=metrics.operational_event_count,
        operational_events=[
            SimAlertEventModel(
                segment_id=event.segment_id,
                start_idx=event.start_idx,
                end_idx=event.end_idx,
                n_candidates=event.n_candidates,
                peak_score=event.peak_score,
            )
            for event in metrics.operational_events
        ],
    )


def _scope(metrics: ScopeMetrics) -> ScopeMetricsModel:
    return ScopeMetricsModel(
        scope=metrics.scope,
        precision=metrics.precision,
        recall=metrics.recall,
        f1=metrics.f1,
        accuracy=metrics.accuracy,
        tn=metrics.tn,
        fp=metrics.fp,
        fn=metrics.fn,
        tp=metrics.tp,
        n_evaluated=metrics.n_evaluated,
        n_anomalous=metrics.n_anomalous,
    )


def _model(row: RowMapping) -> SimModel:
    return SimModel(
        version=cast(str, row["version"]),
        model_key=cast(str, row["model_key"]),
        display_name=cast(str, row["display_name"]),
        score_key=cast(str, row["score_key"]),
        threshold=cast(float, row["threshold"]),
        manifest_sha256=cast(str, row["manifest_sha256"]),
        is_active=bool(row["is_active"]),
    )

File: frontend/src/contracts/simulation.ts
import { z } from 'zod'
import { simDeviceId } from './common'

export const SimModelVersionSchema = z.enum([
  'artifact-lstm-ae-v3',
  'artifact-conv1d-v3',
  'artifact-transformer-v3',
  'artifact-gru-v3',
  'artifact-rnn-v3',
])
export type SimModelVersion = z.infer<typeof SimModelVersionSchema>

export const simModelWindowSizes = Object.freeze({
  'artifact-lstm-ae-v3': 30,
  'artifact-conv1d-v3': 30,
  'artifact-transformer-v3': 30,
  'artifact-gru-v3': 10,
  'artifact-rnn-v3': 10,
} satisfies Readonly<Record<SimModelVersion, number>>)

export const SimModelSchema = z.strictObject({
  version: SimModelVersionSchema,
  model_key: z.string().min(1),
  display_name: z.string().min(1),
  score_key: z.string().min(1),
  threshold: z.number(),
  manifest_sha256: z.string().min(1),
  is_active: z.boolean(),
})
export type SimModel = z.infer<typeof SimModelSchema>

export const SimModelsResponseSchema = z.strictObject({
  request_id: z.string(),
  device_id: z.literal(simDeviceId),
  models: z.array(SimModelSchema),
})
export type SimModelsResponse = z.infer<typeof SimModelsResponseSchema>

export const SetSimActiveModelRequestSchema = z.strictObject({
  model_version: z.string().min(1),
})
export type SetSimActiveModelRequest = z.infer<typeof SetSimActiveModelRequestSchema>

export const SetSimActiveModelResponseSchema = z.strictObject({
  request_id: z.string(),
  device_id: z.literal(simDeviceId),
  active_model_version: z.string().min(1),
})
export type SetSimActiveModelResponse = z.infer<typeof SetSimActiveModelResponseSchema>

export const SimulationScopeMetricsSchema = z
  .strictObject({
    scope: z.enum([
      'timestamp',
      'overlapping_model_windows',
      'non_overlapping_evaluation_bins',
    ]),
    precision: z.number().min(0).max(1),
    recall: z.number().min(0).max(1),
    f1: z.number().min(0).max(1),
    accuracy: z.number().min(0).max(1),
    tn: z.number().int().nonnegative(),
    fp: z.number().int().nonnegative(),
    fn: z.number().int().nonnegative(),
    tp: z.number().int().nonnegative(),
    n_evaluated: z.number().int().nonnegative(),
    n_anomalous: z.number().int().nonnegative(),
  })
  .refine((value) => value.tn + value.fp + value.fn + value.tp === value.n_evaluated, {
    message: 'confusion counts must equal n_evaluated',
    path: ['n_evaluated'],
  })
export type SimulationScopeMetrics = z.infer<typeof SimulationScopeMetricsSchema>

export const SimulationOperationalEventSchema = z
  .strictObject({
    segment_id: z.number().int().nonnegative(),
    start_idx: z.number().int().nonnegative(),
    end_idx: z.number().int().nonnegative(),
    n_candidates: z.number().int().positive(),
    peak_score: z.number(),
  })
  .refine((value) => value.start_idx <= value.end_idx, {
    message: 'start_idx must not be later than end_idx',
    path: ['start_idx'],
  })
export type SimulationOperationalEvent = z.infer<typeof SimulationOperationalEventSchema>

export const SimulationMetricsQuerySchema = z.strictObject({
  modelVersion: SimModelVersionSchema,
  cooldownSamples: z.number().int().min(1).default(10),
})
export type SimulationMetricsQuery = z.input<typeof SimulationMetricsQuerySchema>

const TimestampScopeSchema = SimulationScopeMetricsSchema.safeExtend({
  scope: z.literal('timestamp'),
})
const OverlappingScopeSchema = SimulationScopeMetricsSchema.safeExtend({
  scope: z.literal('overlapping_model_windows'),
})
const BinsScopeSchema = SimulationScopeMetricsSchema.safeExtend({
  scope: z.literal('non_overlapping_evaluation_bins'),
})

export const SimulationMetricsResponseSchema = z
  .strictObject({
    request_id: z.string(),
    device_id: z.literal(simDeviceId),
    model_version: SimModelVersionSchema,
    threshold: z.number(),
    window_size: z.number().int().positive(),
    frame_count: z.number().int().nonnegative(),
    event_count: z.number().int().nonnegative(),
    scored_windows: z.number().int().nonnegative(),
    timestamp_scope: TimestampScopeSchema,
    overlapping_scope: OverlappingScopeSchema,
    bins_scope: BinsScopeSchema,
    operational_event_count: z.number().int().nonnegative(),
    operational_events: z.array(SimulationOperationalEventSchema),
  })
  .refine(
    (value) => value.operational_event_count === value.operational_events.length,
    {
      message: 'operational_event_count must equal operational_events length',
      path: ['operational_event_count'],
    },
  )
export type SimulationMetricsResponse = z.infer<typeof SimulationMetricsResponseSchema>
